imap/message.c

```diff
@@ -69,6 +69,79 @@ static void imap_update_context (IMAP_DATA *idata, int oldmsgcount)
   }
 }
 
+char *imap_header_get_keywords(HEADER *h)
+{
+  if (!HEADER_DATA(h) || !HEADER_DATA(h)->keywords_local)
+    return NULL;
+  return HEADER_DATA(h)->keywords_local;
+}
+
+static int keywords_message(CONTEXT *ctx, HEADER *h, char *new)
+{
+  if (h == NULL)
+    return 0;
+  if (mutt_strcmp (HEADER_DATA(h)->keywords_local, new) == 0)
+    return 0;
+
+  mutt_str_replace (&HEADER_DATA(h)->keywords_local, new);
+  return HEADER_DATA(h)->changed = 1;
+}
+
+int imap_keywords_message(HEADER *h)
+{
+  char buf[LONG_STRING], *new;
+  int i;
+  int changed;
+
+  if (!Context)
+    return 0;
+
+  *buf = '\0';
+  if (HEADER_DATA(h) && HEADER_DATA(h)->keywords_local != NULL) {
+    strncpy(buf, HEADER_DATA(h)->keywords_local, LONG_STRING);
+  }
+
+  if (mutt_get_field("Keywords: ", buf, sizeof(buf), 0) != 0)
+    return 0;
+
+  /* TODO(sileht): ensure we have a valid buf */
+
+  /* each keyword must be atom defined by rfc822 as:
+   *
+   * atom           = 1*<any CHAR except specials, SPACE and CTLs>
+   * CHAR           = ( 0.-127. )
+   * specials       = "(" / ")" / "<" / ">" / "@"
+                    / "," / ";" / ":" / "\" / <">
+		    / "." / "[" / "]"
+   * SPACE          = ( 32. )
+   * CTLS           = ( 0.-31., 127.)
+   *
+   * And must be separated by one space.
+   */
+
+  new = buf;
+  SKIPWS(new);
+  if (*new == '\0')
+    new = NULL;
+
+  changed = 0;
+  if (h != NULL) {
+    changed += keywords_message(Context, h, new);
+  } else {
+#define HDR_OF(index) Context->hdrs[Context->v2r[(index)]]
```

Comment:
@guyzmo as far as i know, clang-format doesn't allow indentation of #define statements.